kernel/src/scan/data_skipping.rs

```diff
@@ -44,22 +44,76 @@ fn get_wide_null_expr(null_col: String) -> Expr {
+fn as_inverted_data_skipping_predicate(expr: &Expr) -> Option<Expr> {
+    use Expr::*;
+    match expr {
+        UnaryOperation { op, expr } => match op {
+            UnaryOperator::Not => as_data_skipping_predicate(expr),
+            UnaryOperator::IsNull => {
+                // to check if a column could NOT have a null, we need two different checks, to see
+                // if the bounds are tight and then to actually do the check
+                if let Column(col) = expr.as_ref() {
+                    let null_col = format!("nullCount.{col}");
+                    Some(Expr::or(
+                        get_tight_not_null_expr(null_col.clone()),
+                        get_wide_not_null_expr(null_col),
+                    ))
+                } else {
+                    // can't check anything other than a col for null
+                    None
+                }
+            }
+        },
+        BinaryOperation { op, left, right } => {
+            let expr = Expr::binary(op.invert()?, left.as_ref().clone(), right.as_ref().clone());
+            as_data_skipping_predicate(&expr)
+        }
+        VariadicOperation { op, exprs } => {
+            let expr = Expr::variadic(op.invert(), exprs.iter().cloned().map(Expr::not));
+            as_data_skipping_predicate(&expr)
+        }
+        _ => None,
+    }
+}
+
 /// Rewrites a predicate to a predicate that can be used to skip files based on their stats.
 /// Returns `None` if the predicate is not eligible for data skipping.
 ///
-/// We normalize each binary operation to a comparison between a column and a literal value
-/// and rewite that in terms of the min/max values of the column.
+/// We normalize each binary operation to a comparison between a column and a literal value and
+/// rewite that in terms of the min/max values of the column.
 /// For example, `1 < a` is rewritten as `minValues.a > 1`.
 ///
-/// Unary `NOT` is transformed recursively then inverted
+/// For Unary `Not`, we push the Not down using De Morgan's Laws to invert everything below the Not.
 ///
-/// Unary `IsNull` checks if the null counts indicate that the column could contain a null
+/// Unary `IsNull` checks if the null counts indicate that the column could contain a null.
 ///
 /// The variadic operations are rewritten as follows:
-/// - `AND` is rewritten as a conjunction of the rewritten operands where we just skip
-///   operands that are not eligible for data skipping.
-/// - `OR` is rewritten only if all operands are eligible for data skipping. Otherwise,
-///   the whole OR expression is dropped.
+/// - `AND` is rewritten as a conjunction of the rewritten operands where we just skip operands that
+///         are not eligible for data skipping.
+/// - `OR` is rewritten only if all operands are eligible for data skipping. Otherwise, the whole OR
+///        expression is dropped.
 fn as_data_skipping_predicate(expr: &Expr) -> Option<Expr> {
     use BinaryOperator::*;
     use Expr::*;
```

Comment:
aside: Looking at the `Equal` and `NotEqual` cases below, I'm a bit surprised they're not consistent with each other? `x = 10` converts to `x <= 10 AND 10 <= x` (which we then extract a data skipping expression for), while `x != 10` converts directly to a data skipping expression? It seems like the former is a simpler/safer approach?

(also, they could have used `Expr::and` and `Expr::or` instead of `and_from` and `or_from`).

Reply:
Yeah, let me see about simplifying this

Reply:
This was acting odd, and I don't have time to fix it right now. I'd like to merge asap since things are "broken", so I've created #250 to remind us to look at this